process VEP_ANNOTATE {
    publishDir params.publish_dir

    input:
      tuple val(patientID), val(sampleID), path(vcfFile)  // as input the seqz files path

    output:
      path("$patientID/$sampleID*")  // the outputs are all the files generated by VEP 
    
    script:

    def cache_version = args!='' && args.cache_version ? "$args.cache_version" : "108"  // cache version
    def assembly = args!='' && args.assembly ? "$args.assembly" : "GRCh38"  // reference assembly
    def ref_genome = args!='' && args.ref_genome ? "$args.ref_genome" : "/scratch/cdslab/ncalonaci/ref_genomes/Homo_sapiens/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38.fasta"  // reference fasta

    """
    vep --input_file "$vcfFile" \
    --vcf \
    --plugin SingleLetterAA \
    --compress_output gzip \
    --offline \
    --cache \
    --dir_cache ~/.vep/ \
    --cache_version $cache_version \
    --assembly $assembly \
    --force \
    --no_stats \
    --everything \
    --use_given_ref \
    --fasta  $ref_genome\
    --fork 23

    """
}
